### PR TITLE
Convert route guards to middleware

### LINKED ISSUE #2234: Replace route guards with route middleware
## Description
In Nuxt 3, we don't have access to Pinia stores anymore, so we should move code from `onBeforeRouteEnter` to the `middleware`, for example, in 
https://github.com/WordPress/openverse/blob/20949acc9063d8dde629cad1c9941b5ad2ca4050/frontend/src/pages/search.vue#L55-L65

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Fixes #2234 by @obulat

## Description

This PR converts the route guards (`beforeRouteEnter` etc) to named and typed middleware. This middleware sets the "back to search" link URL on the single result pages, and the search term on the search pages. The "back to search" link is moved to the search store because `route.meta` does not behave the same way in the `middleware` as it did in the route guards.

## Testing Instructions

There are some navigation tests in the CI that should pass.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: frontend/src/middleware/search.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+export const searchMiddleware: Middleware = ({ redirect, route, $pinia }) => {
+  const {
+    query: { q, searchBy },
+  } = route
+  /**
+   * This anonymous middleware redirects any search without a query to the homepage.
+   * This is meant to block direct access to /search and all sub-routes, with
+   * an exception for the 'creator' filter. The creator filter doesn't send
+   * the search query to the API.
```
Comment (sarayourfriend) [inline frontend/src/middleware/search.ts:13]: > The creator filter doesn't send the search query to the API.

Is this true? How does it work, then? :open_mouth:

File: frontend/src/middleware/search.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { useSearchStore } from "~/stores/search"
+
+import type { Middleware } from "@nuxt/types"
+
+export const searchMiddleware: Middleware = ({ redirect, route, $pinia }) => {
+  const {
+    query: { q, searchBy },
+  } = route
+  /**
+   * This anonymous middleware redirects any search without a query to the homepage.
```
Comment (sarayourfriend) [inline frontend/src/middleware/search.ts:10]: What does "anonymous" mean here? Is that a Nuxt distinction?

**Review by sarayourfriend — COMMENTED** [review]
I am unable to finish reviewing this today, I still need to test it locally. However, the changes look fine to me. I left some questions for clarification of the comments.

File: frontend/src/middleware/search.ts (hunk 2 above)
Comment (obulat) [reply frontend/src/middleware/search.ts:10]: It _was_ anonymous before this change, and now that I've moved it to another file, the comment needs to be updated :) You can see in the changes that previously, `middleware` was a property of the `search.vue` page component.

**Review by sarayourfriend — APPROVED** [review]
LGTM, tested locally and search navigation tests, at least the bits I think would be effected by this change. Specifically I tested going back and forth from results to a single result, using the back to results button and the browser's back button. I also tried to reach search without a query and it correctly redirected to the homepage.

Updating the comments would be helpful but not necessarily blocking.

**Review by zackkrida — APPROVED** [review]
LGTM once the anonymous middleware comment is updated.

File: frontend/src/middleware/search.ts (hunk 1 above)
Comment (obulat) [reply frontend/src/middleware/search.ts:13]: Thank you for noticing this. I realized that this was implemented incorrectly: even if we search by creator (which we currently can only do by manually updating the URL, the UI for searchBy creator was removed), the frontend URL query parameter is `q`: `?q=cat&searchBy=creator` is converted to `?searchBy=creator&creator=cat` API request query. I've removed the check for `searchBy` parameter, and updated the comment accordingly.

**Pull request merged**